### PR TITLE
Implement the wp_cache_*_multiple function in core

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/55029

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by felixarntz** [comment]
@spacedmonkey Happy to review this once #2018 has been committed to core, so that this PR can be refreshed. 👍

File: src/wp-includes/class-wp-comment-query.php [hunk 1]
```diff
@@ -1058,10 +1058,12 @@ protected function fill_descendants( $comments ) {
...
 				}
 
+				$data = array();
 				foreach ( $parent_map as $parent_id => $children ) {
-					$cache_key = "get_comment_child_ids:$parent_id:$key:$last_changed";
-					wp_cache_set( $cache_key, $children, 'comment' );
+					$cache_key          = "get_comment_child_ids:$parent_id:$key:$last_changed";
+					$data[ $cache_key ] = $children;
 				}
+				wp_cache_set_multiple( $data, 'comment' );
```
Comment (adamsilverstein) [inline src/wp-includes/class-wp-comment-query.php:1066]: Nice!

File: src/wp-includes/comment.php [hunk 2]
```diff
@@ -3232,6 +3234,7 @@ function clean_comment_cache( $ids ) {
 		 */
 		do_action( 'clean_comment_cache', $id );
```
Comment (adamsilverstein) [inline src/wp-includes/comment.php]: this now fires before the caches are deleted, maybe move `wp_cache_delete_multiple` call above this loop?

File: src/wp-includes/ms-network.php [hunk 3]
```diff
@@ -110,9 +109,14 @@ function clean_network_cache( $ids ) {
  * @param array $networks Array of network row objects.
  */
 function update_network_cache( $networks ) {
+	if ( ! $networks ) {
```
Comment (adamsilverstein) [inline src/wp-includes/ms-network.php]: why do you need this check? would calling `wp_cache_add_multiple` with an empty array cause problems?

File: src/wp-includes/taxonomy.php [hunk 4]
```diff
@@ -3567,18 +3565,12 @@ function clean_term_cache( $ids, $taxonomy = '', $clean_taxonomy = true ) {
...
 			$taxonomies[] = $term->taxonomy;
 			$ids[]        = $term->term_id;
-			wp_cache_delete( $term->term_id, 'terms' );
 		}
-
+		wp_cache_delete_multiple( $ids, 'terms' );
 		$taxonomies = array_unique( $taxonomies );
 	} else {
+		wp_cache_delete_multiple( $ids, 'terms' );
 		$taxonomies = array( $taxonomy );
```
Comment (adamsilverstein) [inline src/wp-includes/taxonomy.php:3573]: is this `$taxonomies` still used?

File: src/wp-includes/taxonomy.php [hunk 5]
```diff
@@ -3751,11 +3743,16 @@ function update_object_term_cache( $object_ids, $object_type ) {
 		}
 	}
 
+	$cache_values = array();
 	foreach ( $object_terms as $id => $value ) {
 		foreach ( $value as $taxonomy => $terms ) {
-			wp_cache_add( $id, $terms, "{$taxonomy}_relationships" );
+			$cache_values[ $taxonomy ][ $id ] = $terms;
+
```
Comment (adamsilverstein) [inline src/wp-includes/taxonomy.php]: extra whitespace

File: src/wp-includes/taxonomy.php [hunk 6]
```diff
@@ -3751,11 +3743,16 @@ function update_object_term_cache( $object_ids, $object_type ) {
...
 		}
 	}
+	foreach ( $cache_values as $taxonomy => $data ) {
```
Comment (adamsilverstein) [inline src/wp-includes/taxonomy.php:3753]: hopefully all this is already tested, it is challenging to follow

File: src/wp-includes/cache-compat.php [hunk 7]
```diff
@@ -35,3 +35,86 @@ function wp_cache_get_multiple( $keys, $group = '', $force = false ) {
...
+	 * @ignore
+	 * @since 6.0.0
+	 *
+	 * @see wp_cache_delete_multiple()
+	 *
+	 * @param array  $keys  Array of keys under which the cache to deleted.
+	 * @param string $group Optional. Where the cache contents are grouped. Default empty.
+	 * @return array Array of return values organized into groups.
+	 */
+	function wp_cache_delete_multiple( $keys, $group = '' ) {
```
Comment (mitogh) [inline src/wp-includes/cache-compat.php]: Forcing an `array` here removes the need to do any check on the function and relies on the fact that an array is always provided.

```suggestion
	function wp_cache_delete_multiple( array $keys, $group = '' ) {
```

File: src/wp-includes/cache-compat.php [hunk 8]
```diff
@@ -35,3 +35,86 @@ function wp_cache_get_multiple( $keys, $group = '', $force = false ) {
...
+	 *
+	 * @see wp_cache_add_multiple()
+	 *
+	 * @param array  $data  Array of key and value to be set.
+	 * @param string $group Optional. Where the cache contents are grouped. Default empty.
+	 * @param int    $expire Optional. When to expire the cache contents, in seconds.
+	 *                           Default 0 (no expiration).
+	 * @return array Array of return values organized into groups.
+	 */
+	function wp_cache_add_multiple( $data, $group = '', $expire = 0 ) {
```
Comment (mitogh) [inline src/wp-includes/cache-compat.php]: Same as above.

```suggestion
	function wp_cache_add_multiple( array $data, $group = '', $expire = 0 ) {
```

File: src/wp-includes/cache-compat.php [hunk 9]
```diff
@@ -35,3 +35,86 @@ function wp_cache_get_multiple( $keys, $group = '', $force = false ) {
...
+	 *
+	 * @see wp_cache_set_multiple()
+	 *
+	 * @param array  $data  Array of key and value to be set.
+	 * @param string $group Optional. Where the cache contents are grouped. Default empty.
+	 * @param int    $expire Optional. When to expire the cache contents, in seconds.
+	 *                           Default 0 (no expiration).
+	 * @return array Array of return values organized into groups.
+	 */
+	function wp_cache_set_multiple( $data, $group = '', $expire = 0 ) {
```
Comment (mitogh) [inline src/wp-includes/cache-compat.php]: Same as above.

```suggestion
	function wp_cache_set_multiple( array $data, $group = '', $expire = 0 ) {
```

File: src/wp-includes/class-wp-object-cache.php [hunk 10]
```diff
@@ -323,6 +323,67 @@ public function get_multiple( $keys, $group = 'default', $force = false ) {
...
+	/**
+	 * Delete multiple values from the cache in one call.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array  $keys  Array of keys to be delete.
+	 * @param string $group Optional. Where the cache contents are grouped. Default empty.
+	 * @return array Array of return values organized into groups.
+	 */
+	public function delete_multiple( $keys, $group = '' ) {
```
Comment (mitogh) [inline src/wp-includes/class-wp-object-cache.php]: Same as above as well.

```suggestion
	public function delete_multiple( array $keys, $group = '' ) {
```

File: src/wp-includes/class-wp-object-cache.php [hunk 11]
```diff
@@ -323,6 +323,67 @@ public function get_multiple( $keys, $group = 'default', $force = false ) {
...
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array  $data  Array of key and value to be added.
+	 * @param string $group Optional. Where the cache contents are grouped. Default empty.
+	 * @param int    $expire Optional. When to expire the cache contents, in seconds.
+	 *                           Default 0 (no expiration).
+	 * @return array Array of return values organized into groups.
+	 */
+	public function add_multiple( $data, $group = '', $expire = 0 ) {
```
Comment (mitogh) [inline src/wp-includes/class-wp-object-cache.php]: Same as above as well.

```suggestion
	public function add_multiple( array $data, $group = '', $expire = 0 ) {
```

File: src/wp-includes/class-wp-object-cache.php [hunk 12]
```diff
@@ -323,6 +323,67 @@ public function get_multiple( $keys, $group = 'default', $force = false ) {
...
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array  $data  Array of key and value to be set.
+	 * @param string $group Optional. Where the cache contents are grouped. Default empty.
+	 * @param int    $expire Optional. When to expire the cache contents, in seconds.
+	 *                           Default 0 (no expiration).
+	 * @return array Array of return values organized into groups.
+	 */
+	public function set_multiple( $data, $group = '', $expire = 0 ) {
```
Comment (mitogh) [inline src/wp-includes/class-wp-object-cache.php]: Same as above. 

```suggestion
	public function set_multiple( array $data, $group = '', $expire = 0 ) {
```

File: src/wp-includes/taxonomy.php (hunk 4 above)
Comment (spacedmonkey) [reply src/wp-includes/taxonomy.php:3573]: Yeah it is, see below

File: src/wp-includes/cache-compat.php (hunk 7 above)
Comment (spacedmonkey) [reply src/wp-includes/cache-compat.php]: Actioned in https://github.com/WordPress/wordpress-develop/pull/2018

File: src/wp-includes/cache-compat.php (hunk 8 above)
Comment (spacedmonkey) [reply src/wp-includes/cache-compat.php]: Actioned in https://github.com/WordPress/wordpress-develop/pull/2018

File: src/wp-includes/cache-compat.php (hunk 9 above)
Comment (spacedmonkey) [reply src/wp-includes/cache-compat.php]: Actioned in https://github.com/WordPress/wordpress-develop/pull/2018

File: src/wp-includes/class-wp-object-cache.php (hunk 10 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-object-cache.php]: Actioned in https://github.com/WordPress/wordpress-develop/pull/2018

File: src/wp-includes/class-wp-object-cache.php (hunk 11 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-object-cache.php]: Actioned in https://github.com/WordPress/wordpress-develop/pull/2018

File: src/wp-includes/class-wp-object-cache.php (hunk 12 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-object-cache.php]: Actioned in https://github.com/WordPress/wordpress-develop/pull/2018

File: src/wp-includes/comment.php (hunk 2 above)
Comment (spacedmonkey) [reply src/wp-includes/comment.php]: Fixed.

File: src/wp-includes/taxonomy.php (hunk 6 above)
Comment (spacedmonkey) [reply src/wp-includes/taxonomy.php:3753]: I saw unit tests failing here - https://github.com/WordPress/wordpress-develop/runs/5067955057?check_suite_focus=true. 

[This](https://github.com/WordPress/wordpress-develop/pull/2271/commits/7181125d72b28b9e9a3e29bb6c8d97c1054dfa5a) change, fixed it.

**Review by adamsilverstein — APPROVED** [review]
Nice work!

**Pull request closed without merging**